### DISCUSSION TITLE
Shared subscriptions / clean_session

### DISCUSSION THREAD

**Discussion by oscgonfer** [emqx/emqx#13566] [General]
Hi!

So, we have a project were we have small environmental sensors that publish data over MQTT. We use your broker (I have put some questions in the past over here), and we like it a lot, so first of all, thank you for all the great work.

We are seeing an increase on the amount of devices, and we would like to leverage the `Shared subscriptions` features. However, we have been using `clean_session` in the past, for a single subscriber to all the IoT devices that publish data. Our subscriber then passes the data onto a RoR app (repo [here](https://github.com/fablabbcn/smartcitizen-api)). 

Now, the question: together with @timcowlishaw we have been reading various EMQX sources, such as:
- The [docs](https://docs.emqx.com/en/emqx/latest/messaging/mqtt-shared-subscription.html#shared-subscription-and-session)

> The concept of shared subscription and the use of a persistent session in MQTT clients presents a contradiction, making it impossible to use both features simultaneously. If you are using the shared subscription functionality, it is essential to enable the clean session feature on the client by setting the clean_session parameter to true.

- The [blog](https://www.emqx.com/en/blog/mqtt-session) referenced on the docs

- A later entry on the [blog](https://www.emqx.com/en/blog/introduction-to-mqtt5-protocol-shared-subscription)

> It is common to use shared subscriptions together with persistent sessions. However, it's important to note that even if a client in a shared subscription group goes offline, as long as its session and subscription still exist, the MQTT server will continue distributing messages to this session. 
> Considering that the client may be offline for a long time due to failures or other reasons, if the session expiry interval is too long, many messages will be unable to be processed because they are delivered to the offline client.
> A better choice might be to stop considering the subscriber when allocating message load once the subscriber goes offline, even if the session has not expired. Although this behavior is different from normal subscriptions, it is allowed by the MQTT protocol.

Some [discussions](https://github.com/emqx/emqx/discussions/4576), where it seems to be used simultaneously too.

We would like to leverage both features, in principle with just one subscription group, but with various subscribers. At this moment we are not sure about what the official recommendation is, and we have the following questions:
- what is the recommended way to go for these two features?
- if going for it:
    - we assume that each subscriber should reconnect with the a permanent `client_id` to recover the session, and always keep it that way
    - also, consider that healthchecks or ways to monitor if subscribers are up/down are necessary to be able to restart them and avoid message queue overflow. Is that right?
- is it better to keep the session going when the subscriber is offline or to "stop considering the subscriber when allocating message load"?

Thank you so much in advance!

**Comment by zmstone** [comment]
Hi @oscgonfer 

The recommended way in the blogs and old discussions were attempts to provide most common practice which works for both MQTT v3 and v5.

This essentially boils down to how likely EMQX will waste resource for dangling sessions.
The recommended is the simplest way to avoid session `mqueue` overflow.

If your clients are MQTT v3 clients, I would recommend to follow the suggestions. i.e. use clean-session=true (immediately expire the session after the client disconnect).
For MQTT v5, the recommendation would be to use **static client ID** and choose a **short** session expiry interval -- this allows the clients to temporary reconnect and resume the session. 

For EMQX v5, It's worth noting that:
For queued and inflight QoS1 messages, they are re-dispatched to other members in the group when the owning session expires.
Needless to say, but if there is no other member left, then messages are dropped.

**Comment by zmstone** [comment]
Will update the doc.

**Comment by zmstone** [comment]
https://github.com/emqx/emqx-docs/pull/2594

**Comment by oscgonfer** [comment]
aha! thank you so much @zmstone!
@timcowlishaw let's discuss!
